TokboxSubscriber.js

```diff
@@ -0,0 +1,79 @@
+import debug from 'debug'
+import OpenTok from '@opentok/client'
+import { emitLocalEvent, onLocalEvent } from 'libs/localEvents'
+import { autoplayStream, destroyAutoplayStream } from 'libs/audioAutoplay'
+import { tokboxSession } from 'libs/TokboxSession'
+import { getAudioLevelUpdated } from 'libs/TokboxSessionHelpers'
+import { Kit } from 'demio-ui-kit/src'
+
+const { notification } = Kit
+const log = debug('demio:TokboxSessionHelpers')
+
+const subscriberOptions = {
+  insertDefaultUI: false,
+  showControls: false,
+  preferredFrameRate: 30,
+  testNetwork: false,
+  preferredResolution: {
+    width: 1280,
+    height: 720
+  }
+}
+
+export function tokboxSubscribe ({ stream, options: { testNetwork, ...options } = {}, className }) {
+  // @TODO do we still need the fix for duplicated stream??
+  // Probably that was a bug in the previous integration
+  let audioLevel = 0
+
+  const subscription = tokboxSession
+    .subscribe(stream, { ...subscriberOptions, testNetwork, ...options } , (error) => {
+      if (error) {
+        log(error)
+        return console.error(error)
+      }
+    })
+    .on('videoElementCreated', ({ element }) => {
+      const demioVideoElement = autoplayStream({ 
+        stream: subscription.stream, 
+        demioVideoElement: element, 
+        className
+      }) 
+      subscription.demioVideoElement = demioVideoElement
+      emitLocalEvent('TokboxSession:streamCreated', { ...subscription.stream, demioVideoElement, testNetwork })
+      log('Subscriber:videoElementCreated', element)
+    })
+    .on('audioLevelUpdated', (event) => {
+      audioLevel = getAudioLevelUpdated(event, audioLevel, (siglevel) => {
+        if (event.target && event.target.stream && event.target.stream.connection) {
+          const { userId: userID } = JSON.parse(event.target.stream.connection.data) || {}
+          if (userID) {
+            emitLocalEvent('Presenters:audioLevel', {
+              userID,
+              audioLevel: siglevel / 100
+            })
+          }
+        }
+      })
+      // log('Subscriber:audioLevelUpdated', event)
+    })
+    
+  return subscription
+}
+
+// TODO: Use for the layout with shared materials because the webcam size is much smaller
+export function restrictFrameRate (enable) {
+  OpenTok.subscribers
+    .where({ videoType: 'camera' })
+    .forEach((subscriber) => {
+      subscriber.restrictFrameRate(enable)
```

Comment:
`restrictFrameRate` is a very aggressive setting 😄 It will force the subscriber to receive one frame every three seconds.

Are you sure you need such aggressive configuration? What is the use case? Can be achieved by [setPreferredFrameRate](https://tokbox.com/developer/sdks/js/reference/Subscriber.html#setPreferredFrameRate)?

Reply:
You are right, it’s too aggressive. We are not using it.